### ISSUE TITLE
training speed suddenly slow down in benchmark

### ISSUE THREAD

**Issue by crab1234** [horovod/horovod#824]
@alsrgv 
Hi, recently I use horovod to run the benchmark on a real data, and I find that the training speed suddenly slow down. I use 8 GPUs on 2 servers to train the model, at first the training speed can reach near 200 images/sec, then after several steps, the training process get stunned for a long time(during which the GPU capacity drop to 0%) 
the training speed drop like this:
990     images/sec: 195.8 +/- 0.5 (jitter = 15.2)       7.106
990     images/sec: 195.8 +/- 0.5 (jitter = 14.8)       7.209
990     images/sec: 195.8 +/- 0.5 (jitter = 14.8)       7.247
990     images/sec: 195.8 +/- 0.5 (jitter = 14.6)       7.297
990     images/sec: 195.8 +/- 0.5 (jitter = 14.5)       7.042
1000    images/sec: 195.8 +/- 0.5 (jitter = 15.1)       7.050
1000    images/sec: 195.8 +/- 0.5 (jitter = 14.9)       7.049
1000    images/sec: 195.8 +/- 0.5 (jitter = 14.8)       7.132
1000    images/sec: 195.8 +/- 0.5 (jitter = 14.9)       7.437
1000    images/sec: 195.8 +/- 0.5 (jitter = 14.6)       7.213
1000    images/sec: 195.8 +/- 0.5 (jitter = 14.5)       7.054
1000    images/sec: 195.8 +/- 0.5 (jitter = 15.5)       7.069
1000    images/sec: 195.8 +/- 0.5 (jitter = 15.3)       7.211
1010    images/sec: 186.6 +/- 0.5 (jitter = 14.8)       7.268
1010    images/sec: 186.6 +/- 0.5 (jitter = 14.9)       6.983
1010    images/sec: 186.7 +/- 0.5 (jitter = 14.6)       7.091
1010    images/sec: 186.7 +/- 0.5 (jitter = 14.9)       7.332
1010    images/sec: 186.6 +/- 0.5 (jitter = 14.6)       6.854
1010    images/sec: 186.7 +/- 0.6 (jitter = 15.5)       7.067
1010    images/sec: 186.6 +/- 0.6 (jitter = 15.1)       7.260
1010    images/sec: 186.7 +/- 0.6 (jitter = 15.4)       7.375
1020    images/sec: 158.4 +/- 0.7 (jitter = 14.8)       7.218
1020    images/sec: 158.4 +/- 0.7 (jitter = 15.0)       7.296
1020    images/sec: 158.4 +/- 0.7 (jitter = 14.8)       6.996
1020    images/sec: 158.4 +/- 0.7 (jitter = 15.0)       6.595
1020    images/sec: 158.4 +/- 0.7 (jitter = 15.2)       7.133
1020    images/sec: 158.4 +/- 0.7 (jitter = 15.5)       7.372
1020    images/sec: 158.4 +/- 0.7 (jitter = 15.5)       7.058
1020    images/sec: 158.4 +/- 0.7 (jitter = 14.9)       7.175
1030    images/sec: 145.5 +/- 0.8 (jitter = 15.1)       7.020
1030    images/sec: 145.5 +/- 0.8 (jitter = 14.9)       7.074
1030    images/sec: 145.5 +/- 0.8 (jitter = 14.8)       6.964
1030    images/sec: 145.5 +/- 0.8 (jitter = 15.0)       7.057
1030    images/sec: 145.5 +/- 0.8 (jitter = 14.8)       7.186
1030    images/sec: 145.5 +/- 0.8 (jitter = 15.5)       7.044
1030    images/sec: 145.5 +/- 0.8 (jitter = 15.3)       7.163
1030    images/sec: 145.5 +/- 0.8 (jitter = 15.6)       7.025
1040    images/sec: 138.9 +/- 0.8 (jitter = 14.8)       7.175
1040    images/sec: 138.9 +/- 0.8 (jitter = 15.5)       6.993
1040    images/sec: 138.9 +/- 0.8 (jitter = 15.1)       7.156
1040    images/sec: 138.9 +/- 0.8 (jitter = 14.9)       7.137
1040    images/sec: 138.9 +/- 0.8 (jitter = 14.9)       7.135
1040    images/sec: 138.9 +/- 0.8 (jitter = 15.3)       6.977
1040    images/sec: 138.9 +/- 0.8 (jitter = 15.5)       7.075
1040    images/sec: 138.9 +/- 0.8 (jitter = 15.1)       7.242
1050    images/sec: 128.6 +/- 0.8 (jitter = 15.1)       7.078
1050    images/sec: 128.6 +/- 0.8 (jitter = 15.1)       7.237
1050    images/sec: 128.6 +/- 0.8 (jitter = 15.1)       7.097

and I generate the timeline, it seems that after several steps, the NEGOTIATE_ALLREDUCE phase cost much time.

![image](https://user-images.githubusercontent.com/47098550/52698293-2c909a80-2fae-11e9-9ead-72790012f937.png)

I suspect that this may related to a error related to shuffle process.
shuffle process takes a long time, and it will raise warnings like:
"
One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds.This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock.
"
![image](https://user-images.githubusercontent.com/47098550/52700458-dffb8e00-2fb2-11e9-8a68-7305a5ef994c.png)

Does anyone know How can I solve this problem? Thank you very much!

P.S.
my scripts:
mpirun -np 8 -H 3.10.20.6:4,3.10.10.6:4 -bind-to none -map-by slot \
    -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH \
    -mca btl_tcp_if_exclude docker0,lo -x NCCL_SOCKET_IFNAME=^docker0,lo \
    -mca pml ob1 -mca btl ^openib -mca plm_rsh_args "-p 12345" \
    python /share/benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py \
        --model resnet50 \
        --batch_size 64 \
        --variable_update horovod \
        --data_dir /share/ILSVRC2012 \
        --data_name imagenet \
        --num_batches=1000  \
        |& grep -v "Read -1"

and I uses benchmark branch cnn_tf_v1.12_compatible

**Comment by alsrgv** [comment]
@crab1234, can you share what kind of GPUs and CPUs do you have in your servers?  CPU may be underpowered.  Also, @karakusc found that if you set `--datasets_num_private_threads=8` (or some other value), you may have less thread contention which will allow data preprocessing to progress faster.

**Comment by crab1234** [comment]
@alsrgv GPU: nvidia Tesla V100, CPU: Intel(R) Xeon(R) Gold 6152 CPU @ 2.10GHz

This is the performance of CPU and GPU at full speed (200imgs/sec for single GPU)
![fullspeed-cpu](https://user-images.githubusercontent.com/47098550/52766541-6b802800-3062-11e9-83c1-6a4805ea38ac.PNG)

![fullspeed-gpu](https://user-images.githubusercontent.com/47098550/52766547-75a22680-3062-11e9-8a0c-21ad816541ab.PNG)

This is the performance of CPU and GPU when the speed slow down
![lowspeed-cpu](https://user-images.githubusercontent.com/47098550/52766641-bef27600-3062-11e9-811a-3f1395917614.PNG)
![lowspeed-gpu](https://user-images.githubusercontent.com/47098550/52766642-bef27600-3062-11e9-84b7-dd329f7565e1.PNG)

It seems that CPU and GPU is not a problem for this issue.

**Comment by alsrgv** [comment]
@crab1234, could it be that your `/share` is too slow?

**Comment by crab1234** [comment]
@alsrgv  You are right, I found that we use nfs to share data between servers, which is very slow 
now I use ssd to store data in both servers, the training speed is quite fast now. Thank you very much!
By the way, is there any other file sharing system I can use to store training data? It's inconvenient to store data in local disk since every time I have to copy the data to all servers , and it costs a lot of time.

**Comment by alsrgv** [comment]
@crab1234, there are many options from different vendors.  You could check with your IT what are the good options to purchase.
